bin/utest/pgrp.c

```diff
@@ -0,0 +1,137 @@
+#include <assert.h>
+#include <signal.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/wait.h>
+
+#include "utest.h"
+
+int test_setpgid(void) {
+  /* Init process can create and enter new process group. */
+  assert(!setpgid(1, 0));
+
+  pid_t pid = fork();
+  if (pid == 0) {
+    /* Process inherits group of its parent. */
+    assert(getpgid(0) == 1);
+    /* Process can make its own group. */
+    assert(!setpgid(0, 0));
+    /* New group has ID equal to the ID of the moved process. */
+    assert(getpgid(0) == getpid());
+
+    exit(0);
+  }
+  wait(NULL);
+
+  /* It is forbidden to move the process to non-existing group. */
+  assert(setpgid(0, pid));
+  /* It is forbidden to change the group of non-existing process. */
+  assert(setpgid(pid, pid));
+
+  return 0;
+}
+
+static int sig_delivered = 0;
+
+static void sa_handler(int signo) {
+  assert(signo == SIGUSR1);
+  sig_delivered = 1;
+}
+
+static void kill_tests_setup(void) {
+  sig_delivered = 0;
+  setpgid(0, 1);
+
+  sigaction_t newact;
+  sigaction_t oldact;
+
+  newact.sa_handler = sa_handler;
+  sigaction(SIGUSR1, &newact, &oldact);
+}
+
+int test_kill(void) {
+  kill_tests_setup();
+
+  pid_t pid = fork();
+  if (pid == 0) {
+    kill(1, SIGUSR1);
+    /* Signal is not delivered to all processes in the group. */
+    assert(!sig_delivered);
+    exit(0);
+  }
+
+  wait(NULL);
+  /* Signal is delivered to appropriate process. */
+  assert(sig_delivered);
+
+  return 0;
+}
+
+/* In this tests there are five processes marked with: a, b, c, d, e.
+ * Process groups are marked with brackets: (a -> b) --> (c -> d) --> (e)
+ * Process e sends signal to the process group containing c and d.
+ * Process b sends signal to own process group containing a and b. */
+int test_killpg(void) {
```

Comment:
This test looks a little bit complex, but on the other hand it is well commented. I don't have a strong opinion about it, but would be better if you could split it somehow, or come up with a test(s), testing the same functionality but "simpler".

Reply:
I agree with @mzr. If at least one reviewer finds the code difficult to read, it's time to rewrite it.